core/src/main/java/hivemall/smile/tools/MaxEntPredictUDF.java

```diff
@@ -0,0 +1,179 @@
+    private ListObjectInspector featureListOI;
+    private PrimitiveObjectInspector featureElemOI;
+    
+    @Nullable
+    private Vector featuresProbe;
+    private Attribute[] attributes;
+    
+    @Nullable
+    private transient MaxentModel evaluator;
+
+    @Override
+    public ObjectInspector initialize(ObjectInspector[] argOIs) throws UDFArgumentException {
+        if (argOIs.length != 3) {
+            throw new UDFArgumentException("_FUNC_ takes 3");
+        }
+
+        this.modelOI = HiveUtils.asStringOI(argOIs[0]);
+        this.attributesOI = HiveUtils.asStringOI(argOIs[1]);
+        
+        ListObjectInspector listOI = HiveUtils.asListOI(argOIs[2]);
+        this.featureListOI = listOI;
+        ObjectInspector elemOI = listOI.getListElementObjectInspector();
+        if (HiveUtils.isNumberOI(elemOI)) {
+            this.featureElemOI = HiveUtils.asDoubleCompatibleOI(elemOI);
+        }  else {
+            throw new UDFArgumentException(
+                "_FUNC_ takes array<double> for the second argument: "
+                        + listOI.getTypeName());
+        }
+
+        List<String> fieldNames = new ArrayList<String>(2);
+        List<ObjectInspector> fieldOIs = new ArrayList<ObjectInspector>(2);
+        fieldNames.add("value");
+        fieldOIs.add(PrimitiveObjectInspectorFactory.writableStringObjectInspector);
+        fieldNames.add("posteriori");
+        fieldOIs.add(ObjectInspectorFactory.getStandardListObjectInspector(PrimitiveObjectInspectorFactory.writableDoubleObjectInspector));
+        return ObjectInspectorFactory.getStandardStructObjectInspector(fieldNames, fieldOIs);
+    }
+
+    @Override
+    public Object evaluate(@Nonnull DeferredObject[] arguments) throws HiveException {
+        Object arg0 = arguments[0].get();
+        if (arg0 == null) {
+            throw new HiveException("Model was null");
+        }
+        
+        Text model = modelOI.getPrimitiveWritableObject(arg0);
+        try {
+			evaluator = new SepDelimitedTextGISModelReader(model).constructModel();
+		} catch (IOException e) {
+			throw new HiveException(e.getMessage());
+		}
+        
+        Object arg1 = arguments[1].get();
+        if (arg1 == null) {
+            throw new HiveException("arguments were null");
+        }
+        attributes = SmileExtUtils.resolveAttributes(attributesOI.getPrimitiveWritableObject(arg1).toString());
+        
+        Object arg2 = arguments[2].get();
+        if (arg2 == null) {
+            throw new HiveException("array<double> features was null");
+        }
+        this.featuresProbe = parseFeatures(arg2, featuresProbe);
+
+        double[] obs = this.featuresProbe.toArray();
+        
+        String[] names = new String[obs.length];
+  	    float[] values = new float[obs.length];
+	    for (int i = 0; i < obs.length; i++){
+	    	  if (attributes[i].type == AttributeType.NOMINAL){
+	    		  names[i] = i + "_" + String.valueOf(obs[i]).toString();
```

Comment:
`i + "_" + obs[i]` will be perfectly compiled into 
`new StringBuilder().append(i).append("_").append(obj[i])`